File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
     
    	HashMap<Character, Integer> my_map = new HashMap<Character, Integer>();
    	
    	int left_pointer = 0;
    	int max_len = 0;
    	
    	for(int right_pointer = 0; right_pointer < s.length(); right_pointer++) {
    		
    		char current_char = s.charAt(right_pointer);
    		
    		if(my_map.containsKey(current_char) && my_map.get(current_char) >= left_pointer) {
    			
    			left_pointer = my_map.get(current_char) + 1;
    		}

            my_map.put(current_char, right_pointer);
    		max_len = Math.max(max_len, right_pointer - left_pointer + 1);
    	}
    	return max_len;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        left_pointer = 0
        max_len = 0
        my_dict = {}

        for right_pointer in range(len(s)):

            current_char = s[right_pointer]
            
            if current_char in my_dict and my_dict.get(current_char) >= left_pointer:
                left_pointer = my_dict.get(current_char) + 1
            
            my_dict[current_char] = right_pointer

            max_len = max(max_len, right_pointer - left_pointer + 1)
            
        return max_len